### PR TITLE
Bidirectional breadth first search 4e

### PR DESCRIPTION
**Pull request by globalworming**
copied BidirectionalProblem from AIMA3e

### CODE REVIEW

**Comment by globalworming** [comment]
I guess this pull request would be better if I extracted a generic BidirectionalSearch like you did in AIMAe3, right?

**Comment by globalworming** [comment]
The Bidirectional Search can now use different strategies. Unittests are currently failing because there is a mix of expected optimal and not optimal solutions.. will fix that later.

Is the Bidirectional Search ok now?

**Comment by ctjoreilly** [comment]
Hi, I'm afraid it looks like your most recent set of changes broke your unit tests, these should be fixed and you should consider the work complete before pushing to the main repository. Also, if you look at the travis output, see:

https://travis-ci.org/aimacode/aima-java/builds/169782048

the warnings, i.e.:

> /BiDirectionalSearchTest.java:16: warning: NotImplementedException is internal proprietary API and may be removed in a future release
> import sun.reflect.generics.reflectiveObjects.NotImplementedException;
>                                           ^
> /home/travis/build/aimacode/aima-java/test/src/test/java/aima/test/unit/search/support/BiDirectionalSearchTest.java:107: warning: NotImplementedException is internal proprietary API and may be removed in a future release
>        throw new NotImplementedException();

should also be cleaned up. Just throw one of these:

https://docs.oracle.com/javase/7/docs/api/java/lang/UnsupportedOperationException.html

instead.

Best

Ciaran

**Comment by globalworming** [comment]
I am sorry, you are totally right, you should not look into stuff which isn't quite ready.
Changed as requested and wasted a lot of time trying to make parallel A\* optimal. I surrender.. Maybe you have some good ideas? Maybe it's ok that A\* isn't optimal?

File: test/src/test/java/aima/test/unit/search/support/BiDirectionalAStartTest.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+  @Test
+  public void sibiu2bucharest() {
+    StateActionTimeLine timeLine = new StateActionTimeLine();
+    final String start = SIBIU;
+    final String goal = BUCHAREST;
+    final List<GoAction> actual = searchSolution(
+        romaniaRoadMapProblem(start, goal),
+        getHeuristicFunction(goal), getHeuristicFunction(start), timeLine);
+    final List<GoAction> expected;
+    if (actual.contains(new GoAction(FAGARAS))) {
```
Comment (globalworming) [inline test/src/test/java/aima/test/unit/search/support/BiDirectionalAStartTest.java:92]: fixme?
A\* should give an optimal solution. I can't figure out how to make parallel A\* optimal. Is an optimal solution really required?

File: core/src/main/java/aima/core/search/basic/support/StateActionTimeLine.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * Some logging, used for debugging purposes. Todo: one could define some general
+ * search actions (like adding and removing to frontiers) and states (explored states and next in
+ * frontier), for now a message seems enough
+ *
+ * @author wormi
+ */
+
+public class StateActionTimeLine extends ConcurrentLinkedQueue<StateActionTimeLine.MyEntry> {
```
Comment (globalworming) [inline core/src/main/java/aima/core/search/basic/support/StateActionTimeLine.java:14]: if you want, I can just remove that class.

**Comment by ctjoreilly** [comment]
Hi, its going to be another 2-3 weeks before I look into this in-depth but a quick scan looks good.

Thanks

**Comment by globalworming** [comment]
ok, feel free to request changes :)

**Comment by globalworming** [comment]
Hey, any updates on this? I would totally like to do more on this project, but I'd like to see this resolved first, just to get a feeling for how things are done here, if you know what I mean. I would feel more secure about processes, coding conventions etc if I had one pull request approved.

**Comment by ctjoreilly** [comment]
Hey,
Sorry, this is still on my todo list. I am unfortunately spread thin at the moment but am hoping to get this and an official release of the aima3e branch out by end of year.

Best

ciaran

**Comment by ctjoreilly** [comment]
Hi, sorry its been a while since I looked at this. I am running into issues trying to apply this as a patch to the aima4e branch in order to inspect the changes more closely (it appears to not apply the full sequence of changes to particular files, e.g. ProblemFactory.java).  Did you do any force commits to your pull request. Could I ask you to collapse all your changes into a single squashed pull request?

Thanks

Ciaran

**Comment by globalworming** [comment]
no problem.
you'll get a new pull request when I have rebased and squashed

**Pull request closed without merging**